custom_components/zwave_mqtt/climate.py

```diff
@@ -0,0 +1,460 @@
+
+    @property
+    def hvac_action(self):
+        """Return the current running hvac operation if supported.
+
+        Needs to be one of CURRENT_HVAC_*.
+        """
+        return self._hvac_action
+
+    @property
+    def preset_mode(self):
+        """Return preset operation ie. eco, away.
+
+        Needs to be one of PRESET_*.
+        """
+        if not self._mode() or self._preset_mode == PRESET_NONE:
+            return PRESET_NONE
+        return self._hvac_value_label_mapping[self._preset_mode]
+
+    @property
+    def preset_modes(self):
+        """Return the list of available preset operation modes.
+
+        Need to be a subset of PRESET_MODES.
+        """
+        if not self._mode():
+            return []
+        preset_labels = []
+        for value in self._preset_list:
+            preset_labels.append(self._hvac_value_label_mapping[value])
+        preset_labels.append(PRESET_NONE)
+        return preset_labels
+
+    @property
+    def target_temperature(self):
+        """Return the temperature we try to reach."""
+        return self._target_temperature
+
+    @property
+    def target_temperature_low(self) -> Optional[float]:
+        """Return the lowbound target temperature we try to reach."""
+        return self._target_temperature_range[0]
+
+    @property
+    def target_temperature_high(self) -> Optional[float]:
+        """Return the highbound target temperature we try to reach."""
+        return self._target_temperature_range[1]
+
+    async def async_set_temperature(self, **kwargs):
+        """Set new target temperature.
+
+        Must know if single or double setpoint.
+        """
+        current_setpoints = self._current_mode_setpoints()
+        if len(current_setpoints) == 1:
+            setpoint = current_setpoints[0]
+            target_temp = kwargs.get(ATTR_TEMPERATURE)
+            if setpoint is not None and target_temp is not None:
+                setpoint.send_value(target_temp)
+        elif len(current_setpoints) == 2:
+            (setpoint_low, setpoint_high) = current_setpoints
+            target_temp_low = kwargs.get(ATTR_TARGET_TEMP_LOW)
+            target_temp_high = kwargs.get(ATTR_TARGET_TEMP_HIGH)
+            if setpoint_low is not None and target_temp_low is not None:
+                setpoint_low.send_value(target_temp_low)
+            if setpoint_high is not None and target_temp_high is not None:
+                setpoint_high.send_value(target_temp_high)
+
+    async def async_set_fan_mode(self, fan_mode):
+        """Set new target fan mode."""
+        if not self.values.fan_mode:
+            return
+        fan_mode_value = self._fan_label_value_mapping[fan_mode]
+        self.values.fan_mode.send_value(fan_mode_value)
+
+    async def async_set_hvac_mode(self, hvac_mode):
+        """Set new target hvac mode."""
+        if not self._mode():
+            return
+        hvac_mode_value = self._hvac_label_value_mapping.get(hvac_mode.lower())
```

Comment:
`hvac_mode` is a home assistant mode. I don't believe we have set the keys in `self._hvac_label_value_mapping` with home assistant modes, but with z-wave labels.

Reply:
Yes, and I ran into a slight issue with this. The two solutions I can think of are have another dict that maps homeassistant labels to zwave labels, which we then map to zwave values. Second is to ignore any zwave labels that exist for predefined homeassistant modes.

Is there a solution you prefer?

Reply:
Could we have another dict that is almost the reverse of `HVAC_MODE_MAPPINGS`?

Reply:
Yeah, I think that makes sense. It would also let us get rid of ZW_HVAC_MODE_HEAT and ZW_HVAC_MODE_COOL in the process